File: client/src/main/scala/ch/fhnw/ima/bimgur/service/RuntimeService.scala
package ch.fhnw.ima.bimgur
package service

import cats.data.Xor
import ch.fhnw.ima.bimgur.model.activiti._
import io.circe._
import io.circe.generic.auto._
import io.circe.parser._
import org.scalajs.dom.ext.Ajax

import scala.concurrent.Future
import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue

case class RuntimeService(activitiRestUrl: String) {

  val url = activitiRestUrl + "/runtime"

  def getProcessInstances: Future[Xor[Error, Seq[ProcessInstance]]] = {
    val urlWithParams = s"$url/process-instances?includeProcessVariables=true"
    val requestFuture = Ajax.get(urlWithParams)
    requestFuture.map(response => decode[ProcessInstanceList](response.responseText).map(_.data))
  }

}

File: bimgur-activiti-rest-client/src/main/java/ch/fhnw/ima/bimgur/activiti/service/RuntimeService.java
package ch.fhnw.ima.bimgur.activiti.service;

import ch.fhnw.ima.bimgur.activiti.model.ProcessInstance;
import ch.fhnw.ima.bimgur.activiti.model.StartProcessInstanceByIdDto;
import ch.fhnw.ima.bimgur.activiti.model.StartProcessInstanceByKeyDto;
import ch.fhnw.ima.bimgur.activiti.model.util.ResultList;
import ch.fhnw.ima.bimgur.activiti.service.util.ResultListExtractor;
import io.reactivex.Observable;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.POST;

public interface RuntimeService {

    @GET("runtime/process-instances")
    Observable<ResultList<ProcessInstance>> getProcessInstancesResultList();

    default Observable<ProcessInstance> getProcessInstances() {
        return ResultListExtractor.extract(getProcessInstancesResultList());
    }

    @Headers("Content-type: application/json")
    @POST("runtime/process-instances")
    Observable<ProcessInstance> startProcessInstance(@Body StartProcessInstanceByKeyDto startData);

    @Headers("Content-type: application/json")
    @POST("runtime/process-instances")
    Observable<ProcessInstance> startProcessInstance(@Body StartProcessInstanceByIdDto startData);

}